### PR TITLE
[SPARK-15103][SQL] Refactored FileCatalog class to allow StreamFileCatalog to infer partitioning

### PR DESCRIPTION
**Pull request by tdas**
## What changes were proposed in this pull request?

File Stream Sink writes the list of written files in a metadata log. StreamFileCatalog reads the list of the files for processing. However StreamFileCatalog does not infer partitioning like HDFSFileCatalog.

This PR enables that by refactoring HDFSFileCatalog to create an abstract class PartitioningAwareFileCatalog, that has all the functionality to infer partitions from a list of leaf files. 
- HDFSFileCatalog has been renamed to ListingFileCatalog and it extends PartitioningAwareFileCatalog by providing a list of leaf files from recursive directory scanning. 
- StreamFileCatalog has been renamed to MetadataLogFileCatalog and it extends PartitioningAwareFileCatalog by providing a list of leaf files from the metadata log.
- The above two classes has been moved into their own files as they are not interfaces that should be in fileSourceInterfaces.scala.
## How was this patch tested?
- FileStreamSinkSuite was update to see if partitioning gets inferred, and on reading whether the partitions get pruned correctly based on the query. 
- Other unit tests are unchanged and pass as expected.

### CODE REVIEW

**Comment by tdas** [comment]
@marmbrus

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/fileSourceInterfaces.scala [hunk 1]
```diff
@@ -331,7 +319,44 @@ class HDFSFileCatalog(
     }
   }
 
-  protected def prunePartitions(
+  override def allFiles(): Seq[FileStatus] = leafFiles.values.toSeq
+
+  override def getStatus(path: Path): Array[FileStatus] = leafDirToChildrenFiles(path)
+
+  protected def inferPartitioning(): PartitionSpec = {
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/fileSourceInterfaces.scala]: @marmbrus These methods are basically unchanged, just moved around from HDFSFileCatalog to PartitioningAwareFileCatalog.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/fileSourceInterfaces.scala [hunk 2]
```diff
@@ -365,11 +390,78 @@ class HDFSFileCatalog(
...
+/**
+ * A file catalog that caches metadata gathered by scanning all the files present in `paths`
+ * recursively.
+ *
+ * @param parameters as set of options to control discovery
+ * @param paths a list of paths to scan
+ * @param partitionSchema an optional partition schema that will be use to provide types for the
+ *                        discovered partitions
+ */
+class HDFSFileCatalog(
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/fileSourceInterfaces.scala]: existing: can you move the implementation into their own files (each separate).  they aren't really interfaces.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MetadataLogFileCatalog.scala [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.execution.datasources._
+
+
+/**
+ * A [[FileCatalog]] that generates the list of files to processing by reading them from the
+ * metadata log files generated by the [[FileStreamSink]].
+ */
+class MetadataLogFileCatalog(sparkSession: SparkSession, path: Path)
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MetadataLogFileCatalog.scala]: This is basically StreamFileCatalog renamed. Github not showing it as a rename.

**Comment by tdas** [comment]
ping @marmbrus

**Comment by marmbrus** [comment]
LGTM

**Comment by tdas** [comment]
Thanks! Merging this to master and 2.0

**Pull request closed without merging**